Fixed #27334 -- TemporaryUploadedFile moved to destination when possible

achidlow opened the pull request:
When a FileField is set to an instance of File that is not also an
instance of FieldFile, pre_save will pass that object as the contents to
Storage.save().

As a result, when an uploaded file is written to temporary storage,
if possible it will be moved (rather than copied) to the upload
destination.

All tests pass under Linux & Windows. The tests will pass without the changes to files.py on Windows however, because Windows does not support moving open files. On Linux they will fail without the changes to files.py, thus verifying that the test is actually testing something.

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
The test looks overcomplicated -- I don't think the request/response cycle is required to test this fix. Since only `FileField` is modified, maybe `tests/model_fields/test_filefield.py` would be a better place.

achidlow commented:
Thanks for the feedback, I've rewritten the test at the level you suggested and simplified as much as possible, and squashed the commits in my branch.

charettes commented:
I wonder if we should also test the correct permissions are set on the moved file and that the file is actually copied when using a non-local file storage.

achidlow commented:
@charettes there are permissions tests in [file_strorage/tests.py](https://github.com/django/django/blob/ac1975b18b5a33234284bec86e5a5bb44a4af976/tests/file_storage/tests.py#L922), although they're using stream-file objects so the move-if-temporary-file logic won't be exercised. I could simply duplicate that test class using `TemporaryUploadedFile` as the content.

As for copying when non-local file storage, I'm not 100% sure what you mean by non-local. Would something like the tests linked above (ie copying from in-memory contents) count as non-local storage?

charettes commented:
@achidlow I think both of my points are moot, it looks like `FileSystemStorage` [apply permission flags independently of whether or not the file was created or moved](https://github.com/django/django/blob/a4e9e834e3dfc8d5a024a78c765f193105d41a48/django/core/files/storage.py#L373-L374) and Django doesn't ship with non-local storages (e.g. S3) by default.

achidlow commented:
@timgraham just wondering if you're happy with the updated tests?

timgraham commented:
merged in f734e2d4b2fc4391a4d097b80357724815c1d414, thanks!
